### PR TITLE
Add configurable workspace prefix for S3 Backend

### LINKED ISSUE #13184: Allow configurable `keyEnvPrefix`
### Terraform Version
0.9.2

### Affected Resource(s)
remote state backend (s3)

### Terraform Configuration Files
```hcl
terraform {
	required_version = ">= 0.9.2"
	backend "s3" {
		bucket = "platform-infrastructure"
		key = "environments/alpha/master.tfstate"
		region = "us-west-2"
	}
}
```

### Expected Behavior
I expect not to be forced into some hand-held path by a tool (terraform) for advanced users (anybody who uses terraform).

Read: let me configure `keyEnvPrefix`. I've already created my own bucket, and nested my environment files under `/environments/{environment}/master.tfstate`

### Actual Behavior
My states are stored and pulled from `platform-infrastructure/env:/alpha/environments/alpha/master.tfstate

The first `alpha` originating from terraform env, and the second `alpha` comes from my own (legacy) namespacing. (We're migrating from self-managed environments to TF managed)

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Try to use terraform with an S3 backend

### PR DESCRIPTION
**Pull request by rowleyaj**
Fixes #13184, addresses part of #15358

@jbardin @apparentlymart this is my first foray into the world of Go so go please go easy :)

Does this address the points raised in the other issue? The tests still pass although I'm not sure how to add some specifically related to this. I've also tested it locally and it successfully migrated my dummy setup from the old state structure to the new.

```
$ tree
.
├── env: # Old structure
│   ├── testing
│   │   └── v1
│   │       └── terraform-remote-state-example
│   ├── testing2
│       └── v1
│           └── terraform-remote-state-example
├── terraform-state # New Structure
│   ├── testing
│   │   └── v1
│   │       └── terraform-remote-state-example
│   └── testing2
│       └── v1
│           └── terraform-remote-state-example
└── v1 # Default workspace
    └── terraform-remote-state-example
```

```
terraform {
  required_version = "~> 0.10"

  backend "s3" {
    bucket  = "rowleyaj-tf-state-demo"
    key     = "v1/terraform-remote-state-example"
    region  = "us-east-1"
    encrypt = true

    dynamodb_table = "rowleyaj-terraform-state-lock"
    workspace_key_prefix = "terraform-state"
  }
}
```

### CODE REVIEW

**Comment by rowleyaj** [comment]
Not sure why tests are failing on this but I don't think it's related, any help appreciated:
```
--- FAIL: TestResourceProvider_stop (2.05s)
	resource_provisioner_test.go:80: should finish
FAIL
FAIL	github.com/hashicorp/terraform/builtin/provisioners/local-exec	4.192s
```

**Comment by rowleyaj** [comment]
Tests pass now, was a strange timeout :)

File: website/docs/backends/types/s3.html.md [hunk 1]
```diff
@@ -108,3 +108,5 @@ The following configuration options or environment variables are supported:
  * `assume_role_policy` - (Optional) The permissions applied when assuming a role.
  * `external_id` - (Optional) The external ID to use when assuming the role.
  * `session_name` - (Optional) The session name to use when assuming the role.
+ * `workspace_key_prefix` - (Optional) The prefix applied to the state path inside the bucket.
+   This is only relevant when using a non-default workspace.
```
Comment (josephholsten) [inline website/docs/backends/types/s3.html.md]: should we mention the default is `"env:"`?

File: backend/remote-state/s3/backend.go [hunk 2]
```diff
@@ -139,6 +139,13 @@ func New() backend.Backend {
 				Description: "The permissions applied when assuming a role.",
 				Default:     "",
 			},
+
+			"workspace_key_prefix": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The prefix applied to the state path inside the bucket",
```
Comment (jbardin) [inline backend/remote-state/s3/backend.go]: Can we mention that it's for non-default workspaces here? Maybe:

    applied to the non-default state path

**Pull request merged**